static/app/views/explore/toolbar/toolbarGroupBy.tsx

```diff
@@ -60,61 +74,203 @@ export function ToolbarGroupBy({disabled}: ToolbarGroupByProps) {
+    insertColumn,
+    updateColumnAtIndex,
+    deleteColumnAtIndex,
+    swapColumnsAtIndex,
+  } = useDragNDropColumns({columns: groupBys});
 
-  const deleteGroupBy = useCallback(
-    (index: number) => {
-      const newGroupBys = groupBys.filter((_, orgIndex) => index !== orgIndex);
-      setGroupBys?.(newGroupBys);
-    },
-    [setGroupBys, groupBys]
-  );
+  useEffect(() => {
+    setGroupBys(editableColumns.map(({column}) => column ?? ''));
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [editableColumns]);
 
   return (
     <ToolbarSection data-test-id="section-group-by">
-      <ToolbarHeader>
+      <StyledToolbarHeader>
         <ToolbarLabel disabled={disabled}>{t('Group By')}</ToolbarLabel>
         <ToolbarHeaderButton
           disabled={disabled}
           size="zero"
-          onClick={addGroupBy}
+          onClick={insertColumn}
           borderless
           aria-label={t('Add Group')}
           icon={<IconAdd />}
         />
-      </ToolbarHeader>
+      </StyledToolbarHeader>
       <div>
-        {groupBys.map((groupBy, index) => (
-          <ToolbarRow key={index}>
-            <CompactSelect
-              searchable
-              disabled={disabled}
-              options={options}
-              value={groupBy}
-              onChange={newGroupBy => setGroupBy(index, newGroupBy)}
-            />
-            <Button
-              borderless
-              icon={<IconDelete />}
-              size="zero"
-              disabled={disabled || (groupBys.length <= 1 && groupBy === '')}
-              onClick={() => deleteGroupBy(index)}
-              aria-label={t('Remove Group')}
-            />
-          </ToolbarRow>
-        ))}
+        <ColumnEditor
+          columns={editableColumns}
+          onColumnChange={updateColumnAtIndex}
+          onColumnDelete={deleteColumnAtIndex}
+          onColumnSwap={swapColumnsAtIndex}
+          options={options}
+        />
       </div>
     </ToolbarSection>
   );
 }
+
+const StyledToolbarHeader = styled(ToolbarHeader)`
+  margin-bottom: ${space(1)};
+`;
+
+interface ColumnEditorProps {
+  columns: Column[];
+  onColumnChange: (i: number, column: string) => void;
+  onColumnDelete: (i: number) => void;
+  onColumnSwap: (i: number, j: number) => void;
+  options: SelectOption<string>[];
+}
+
+export function ColumnEditor({
+  columns,
+  onColumnChange,
+  onColumnDelete,
+  onColumnSwap,
+  options,
+}: ColumnEditorProps) {
+  const [resultMode] = useResultMode();
+
+  const sensors = useSensors(
+    useSensor(PointerSensor),
+    useSensor(KeyboardSensor, {
+      coordinateGetter: sortableKeyboardCoordinates,
+    })
+  );
+
+  function handleDragEnd(event) {
+    const {active, over} = event;
+
+    if (active.id !== over.id) {
+      const oldIndex = columns.findIndex(({id}) => id === active.id);
+      const newIndex = columns.findIndex(({id}) => id === over.id);
+      onColumnSwap(oldIndex, newIndex);
+    }
+  }
```

Comment:
This feels like it should be handled by the `useDragNDropColumns` hook as it deals with the id which is injected by it.